### PR TITLE
[dLLM] Run several denoise forwards per scheduled FDFO round

### PR DESCRIPTION
**Pull request by wenxuewuhd**
## Motivation

Under FDFO, one scheduler round is one denoise forward. Everything the round does around that forward — committing resolved blocks, admitting waiting requests, processing results, rebuilding the batch — is paid once per forward. dLLM also disables the overlap scheduler (each denoise step has a hard input dependency on the previous step's result), so that host work is exposed directly as device idle rather than hidden behind the next forward.

At large batch this is a real share of the step: the forward itself is the only part that scales with the model, while the round overhead is roughly fixed.

## Modifications

`SGLANG_DLLM_FDFO_STEPS_PER_ROUND` (new, default `1` = today's behavior) makes a round issue N denoise forwards instead of one. The batch is frozen for the inner steps: rows that resolve mid-round freeze themselves inside the step, and commits, admissions and result processing all happen once at the end of the round. The per-round host work is therefore amortized over N forwards.

Two hooks are added to `DllmAlgorithm` so an algorithm can keep its per-request state in batched device tensors for the whole round (`fdfo_batched_begin` / `fdfo_batched_step` / `fdfo_batched_all_finished`), which is what makes an inner step cost no host gather/scatter. `JointThreshold` implements them; an algorithm that returns `None` from `fdfo_batched_begin` gets a plain loop with the same semantics, which is the path non-NPU dLLM runs today.

The round still exits early once every row has resolved, so a short round does not issue no-op forwards. That probe costs one scalar sync per inner step and was measured worth keeping at every batch size tried — skipping it on the theory that a wide batch never resolves mid-round made the tail of a run slower, not faster.

### Trade-offs, stated plainly

- **Emission latency.** A row's accepted tokens are published at the end of the round, so up to N-1 forwards later than they would have been. With N=2 that is one extra forward of delay.
- **A few wasted forwards.** Rows that resolve on inner step 1 are re-forwarded (frozen) until the round ends.
- **N has no universal value.** See Speed Tests.

### Blast radius

| File | Scope |
| --- | --- |
| `dllm/algorithm/base.py`, `joint_threshold.py` | dLLM only — no non-dLLM callers; the whole inner loop is behind `fdfo_steps_per_round > 1`, which is `1` unless the env var is set. |
| `environ.py` | One `EnvInt(1)` descriptor; the default reproduces today exactly. |

One device-specific site, in the round loop:

```python
if inner_steps and _is_npu:
    forward_batch.mark_forward_metadata_ready()
```

Attention metadata is stable across a round's denoise steps because the first forward already planned it, so NPU marks it ready once and the inner forwards skip re-planning. Other platforms re-plan every inner step: correct, one optimization short. The same pattern already exists upstream in `_run_sync`, so this is not a new device assumption. Everything else — the loop, the early exit, the batched-state hooks — is device-agnostic.

## Accuracy Tests

gsm8k (zero-shot, 256 questions, parallel 128) on LLaDA2.1-mini / Ascend 910B3, across three server launches:

| Config | Runs | Accuracy |
| --- | --- | --- |
| N=1 (default) | 4, over 2 server launches | 0.828, 0.832, 0.840, 0.840 |
| N=2 | 2, one launch | 0.855, 0.863 |

No regression. The small upward difference for N=2 is **not** claimed as a benefit: server-side generation is nondeterministic run to run under concurrency regardless of this flag, the two N=1 launches alone differ by 1.2 points, and two samples on one launch is not enough to attribute the remaining gap to N. What *is* checked, in the setting where it is checkable, is stronger — see Unit Tests: an N-step round produces token-identical output to N single-step rounds.

## Speed Tests and Profiling

Same workload, measured as wall-clock time to complete the fixed 256-question set:

| Config | Runs | Completion time |
| --- | --- | --- |
| N=1 (default) | 4 | 23.205, 22.418, 23.957, 21.951 s |
| N=2 | 2 | 21.028, 20.385 s |

Mean 22.88 s -> 20.71 s, **10.5% faster**. The N=1 spread across two launches is 9.1%, which is why four baseline runs were taken rather than one: even against that spread the two sets do not overlap — the slowest N=2 run (21.028 s) beats the fastest N=1 run (21.951 s).

This is completion time for a fixed question set, not steady-state throughput. It includes batch ramp-up and drain, where the amortization has less to work with, so steady-state gain at a saturated batch should be at least this.

**N has no universal value, and this PR does not recommend one.** Measurements of this knob disagree across workloads and across observers — an earlier internal run put it at +3.4% on gsm8k and +11.3% on long sequences, and operators of this model report the opposite ordering. That is consistent with the mechanism: the gain is `per-round host work / (per-round host work + N x forward time)`, so it depends on how expensive the round overhead is relative to a forward for a given model, sequence length and batch size. Sweep it for the payload rather than adopting a number. The default stays 1.

## Unit Tests

`test/registered/unit/dllm/test_fdfo_multi_step_round.py` (CPU-only, no server, `base-a-test-cpu`), 3 cases against a scripted model runner:

- an N-step round leaves exactly what N single-step rounds leave — same forwards, same tokens (`torch.equal`), same accept lengths — asserted for the batched-state path;
- the same contract for the fallback path where `fdfo_batched_begin` returns `None`, which is the loop non-NPU dLLM runs;
- the early exit: a block that resolves on the first step must not issue the remaining inner forwards.

The first two also assert the window actually exercises unresolved carry-over, so the comparison cannot pass vacuously. Mutation-checked: an off-by-one in `inner_steps` fails 2 of the 3, removing the early exit fails 1.

## Configuration

| Value | Behavior |
| --- | --- |
| `1` (default) | Today's behavior: one denoise forward per scheduler round. |
| `N > 1` | N forwards per round, batch frozen for the inner steps. |

## Relationship to the other dLLM PRs

Independent of #33161 (denoise logits) and the mixed-round PR — no shared code path, and all three were trial-merged against each other cleanly. The only file more than one of them touches is `environ.py`, and they add descriptors at different anchors.

One detail worth knowing on rebase: the unit test pins the portable denoise path with `patch(..., "_argmax_softmax_prob_fused", None, create=True)`. That hook is introduced by #33161; `create=True` is what lets this test hold both before and after that PR lands. Verified passing in both configurations.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

### CODE REVIEW

File: test/registered/unit/dllm/test_fdfo_multi_step_round.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+def _algorithm(steps_per_round: int, *, vectorized: bool, max_post_edit_steps: int):
+    config = DllmConfig(
+        algorithm="JointThreshold",
+        algorithm_config={
+            "threshold": 0.5,
+            # p is a probability, so an edit threshold above 1 disables T2T and
+            # leaves a pure mask-to-token trajectory.
+            "edit_threshold": 2.0,
+            "max_post_edit_steps": max_post_edit_steps,
```
Comment (rwang5203) [inline test/registered/unit/dllm/test_fdfo_multi_step_round.py]: These tests disable T2T, so they miss cross-round `post_edit_steps` state.

Add a K=1-versus-K T2T case that carries and exhausts the edit budget.

Comment (wenxuewuhd) [reply test/registered/unit/dllm/test_fdfo_multi_step_round.py]: Right — `edit_threshold: 2.0` disabled T2T, and with masks still in the block `no_mask_active` never fires either, so `post_edit_steps` stayed 0 in every case.

Added a scripted K=1 / K=2 / K=4 case on both the batched and fallback paths: forward 1 unmasks the block, forwards 2 and 3 edit it via T2T, forward 4 exhausts the budget (`max_post_edit_steps=2`) and finishes the row. The final token is the assertion — a budget re-initialised per round would leave K=2 with headroom at forward 4 and edit one step further.

Zeroing `post_edit_steps` in `fdfo_batched_begin`, or dropping the write-back in `fdfo_batched_end`, turns only this case red; the pre-existing three stay green, which is the gap you pointed at.

**Comment by rwang5203** [comment]
The body describes mixed batching and `SGLANG_ENABLE_DLLM_MIXED_BATCH`, while this head implements multi-forward rounds behind `SGLANG_DLLM_FDFO_STEPS_PER_ROUND`.

Replace it with this change's rationale, configuration, test plan, and measurements.

**Comment by wenxuewuhd** [comment]
> frozen

Replaced — the body now describes this change: the rationale (round host work exposed as device idle since dLLM can't overlap), the SGLANG_DLLM_FDFO_STEPS_PER_ROUND configuration, the unit-test plan, and gsm8k accuracy plus completion-time measurements on 910B3.
